File: LeetCode/CPP/LeetCode131.cpp
// Given a string s, partition s such that every substring of the partition is a palindrome.
// Return all possible palindrome partitioning of s.

// Look For Reference In the Book On Pg No 102-103


#include<iostream>
#include<vector>

using namespace std;

bool isPalindrome(string str, int start, int end){
    while(start<end){
        if(str[start]!=str[end])    return false;
        start++;
        end--;
    }
    return true;
}

void findPartitions(int start, string str, vector<string>& partition, vector<vector<string>>& allPartition, int n){
    if(start == n){
        allPartition.push_back(partition);
        return;
    }
    for(int end=start;end<n;end++){
        if(isPalindrome(str, start, end)){
            partition.push_back(str.substr(start, end-start+1));
            findPartitions(end+1, str, partition, allPartition, n);
            partition.pop_back();
        }
    }
}

vector<vector<string>> findAllPartition(string str){
    vector<vector<string>> allPartitions;
    vector<string> partition;
    int n = str.size();
    findPartitions(0, str, partition, allPartitions, n);
    return allPartitions;
}

int main(){
    string str = "aab";
    vector<vector<string>> res = findAllPartition(str);
    cout<<"[";
    for(auto &part : res){
        cout<<"[";
        for(string pal : part){
            cout<<"'"<<pal<<"'";
        }
        cout<<"]";
    }
    cout<<"]";
    return 0;
}

File: LeetCode/JAVA/LeetCode131.java
// Given a string s, partition s such that every substring of the partition is a palindrome.
// Return all possible palindrome partitioning of s.

// Look For Reference In the Book On Pg No 102-103



import java.util.ArrayList;
import java.util.List;

public class LeetCode131{
    public static boolean isPalidrome(String str, int start, int end){
        while(start<end){
            if(str.charAt(start)!=str.charAt(end)){
                return false;
            }
            start++;
            end--;
        }
        return true;
    }
    public static void findPartitions(int start, String str, List<String> partition, List<List<String>> allPartitions){
        if(start==str.length()){
            allPartitions.add(new ArrayList<>(partition));
            return;
        }
        for(int end = start;end<str.length();end++){
            if(isPalidrome(str, start, end)){
                partition.add(str.substring(start, end+1));
                findPartitions(end+1, str, partition, allPartitions);
                partition.remove(partition.size()-1);
            }
        }
    }
    public static List<List<String>> palindrome_partitioning(String str){
        List<List<String>> allPartitions = new ArrayList<>();
        findPartitions(0, str, new ArrayList<>(), allPartitions);
        return allPartitions;
    }
    public static void main(String[] args) {
        String str = "suyash";
        List<List<String>> res = palindrome_partitioning(str);
        System.out.println(res);
    }
}